File: reviews/2026-08-25-post-ruling-trials/go-implementation/erfval/narrative.go
package main

import (
	"strings"
	"unicode"
)

// ---------------------------------------------------------------------------
// SCOPE AREA 2: narrative bindings.
// ERF-31 (grammar, recognition, anchor), ERF-32 (staleness), ERF-33
// (unresolved id), ERF-34 (the narrative document), ERF-47 (staleness rule).
//
// The grammar as given in ERF-31:
//
//   narrative-binding ::= "<!--" ws* "claims:" ws+ ids ws+ anchor
//                         ws+ "bound-at=" date ws* "-->"
//   date     ::= YYYY "-" MM "-" DD
//   ids      ::= id (ws+ id)*
//   id       ::= one or more characters, none of them whitespace or '"'
//   anchor   ::= '"' char* '"'
//   char     ::= any character other than '"' and '\', or one of the
//                two-character escapes '\"' and '\\'
//
// `ws` is never defined. This implementation takes it to be any Unicode
// whitespace, which lets a binding span lines. See ambiguities.md A-28.
// ---------------------------------------------------------------------------

type Binding struct {
	File     string
	Raw      string // the whole comment, `<!--` .. `-->`
	Start    int    // byte offset of `<!--` in the body
	End      int    // byte offset just past `-->`
	Line     int
	IDs      []string
	Anchor   string // unescaped
	BoundAt  string
	ParseErr string // non-empty if the comment failed the grammar
	Passage  string
}

// findComments returns every HTML comment in the body, using the HTML rule
// that a comment ends at the first `-->`.
func findComments(body string) []struct{ start, end int } {
	var out []struct{ start, end int }
	i := 0
	for {
		s := strings.Index(body[i:], "<!--")
		if s < 0 {
			return out
		}
		s += i
		e := strings.Index(body[s+4:], "-->")
		if e < 0 {
			// An unterminated comment. Everything to end of file is the
			// comment under the HTML rules.
			out = append(out, struct{ start, end int }{s, len(body)})
			return out
		}
		e = s + 4 + e + 3
		out = append(out, struct{ start, end int }{s, e})
		i = e
	}
}

// isRecognizedBinding implements ERF-31's recognition rule, which is
// deliberately looser than the grammar: "A comment opening `<!--` followed by
// `claims:` IS a narrative binding: recognizing one and validating one are
// separate acts, and a consumer performs them in that order."
func isRecognizedBinding(comment string) bool {
	if !strings.HasPrefix(comment, "<!--") {
		return false
	}
	rest := strings.TrimLeftFunc(comment[4:], unicode.IsSpace)
	return strings.HasPrefix(rest, "claims:")
}

// parseBinding validates a recognized comment against the ERF-31 grammar.
func parseBinding(comment string) *Binding {
	b := &Binding{Raw: comment}

	if !strings.HasSuffix(comment, "-->") {
		b.ParseErr = "comment is not terminated by `-->`"
		return b
	}
	inner := comment[4 : len(comment)-3]

	// "<!--" ws* "claims:"
	p := 0
	for p < len(inner) && isWS(rune(inner[p])) {
		p++
	}
	if !strings.HasPrefix(inner[p:], "claims:") {
		b.ParseErr = "does not open `claims:`"
		return b
	}
	p += len("claims:")

	// ws+ before the first id
	n := skipWS(inner, p)
	if n == p {
		b.ParseErr = "no whitespace after `claims:` (grammar requires ws+)"
		return b
	}
	p = n

	// ids ::= id (ws+ id)*, terminated by the anchor's opening quote.
	// An id may contain no whitespace and no '"', so the first '"' after the
	// ids is unambiguously the anchor's opening delimiter.
	for {
		if p >= len(inner) {
			b.ParseErr = "ends after the ids: no anchor and no `bound-at=`"
			return b
		}
		if inner[p] == '"' {
			break
		}
		start := p
		for p < len(inner) && !isWS(rune(inner[p])) && inner[p] != '"' {
			p++
		}
		if p == start {
			b.ParseErr = "empty id token"
			return b
		}
		b.IDs = append(b.IDs, inner[start:p])
		n := skipWS(inner, p)
		if n == p && p < len(inner) && inner[p] != '"' {
			b.ParseErr = "id runs straight into the anchor with no whitespace between"
			return b
		}
		if n == p && p < len(inner) && inner[p] == '"' {
			b.ParseErr = "no whitespace between the last id and the anchor (grammar requires ws+)"
			return b
		}
		p = n
	}
	if len(b.IDs) == 0 {
		b.ParseErr = "no claim ids before the anchor (grammar requires at least one)"
		return b
	}

	// anchor ::= '"' char* '"'
	p++ // consume opening quote
	var sb strings.Builder
	closed := false
	for p < len(inner) {
		c := inner[p]
		if c == '"' {
			closed = true
			p++
			break
		}
		if c == '\\' {
			if p+1 >= len(inner) {
				b.ParseErr = "anchor ends with a dangling backslash"
				return b
			}
			nx := inner[p+1]
			if nx != '"' && nx != '\\' {
				b.ParseErr = "anchor contains `\\" + string(nx) + "`; the only escapes the grammar defines are `\\\"` and `\\\\`"
				return b
			}
			sb.WriteByte(nx)
			p += 2
			continue
		}
		sb.WriteByte(c)
		p++
	}
	if !closed {
		b.ParseErr = "anchor is not closed by an unescaped `\"`"
		return b
	}
	b.Anchor = sb.String()

	// ws+ "bound-at=" date ws* (end)
	n = skipWS(inner, p)
	if n == p {
		b.ParseErr = "no whitespace between the anchor and `bound-at=` (grammar requires ws+)"
		return b
	}
	p = n
	if !strings.HasPrefix(inner[p:], "bound-at=") {
		b.ParseErr = "expected `bound-at=` after the anchor, found " + shortRest(inner[p:])
		return b
	}
	p += len("bound-at=")
	dstart := p
	for p < len(inner) && !isWS(rune(inner[p])) {
		p++
	}
	b.BoundAt = inner[dstart:p]
	if !reDate.MatchString(b.BoundAt) {
		b.ParseErr = "`bound-at=" + truncate(b.BoundAt, 30) + "` is not a YYYY-MM-DD date"
		return b
	}
	// ws* then end of the comment
	p = skipWS(inner, p)
	if p != len(inner) {
		b.ParseErr = "trailing content after the date: " + shortRest(inner[p:])
		return b
	}
	return b
}

func isWS(r rune) bool { return unicode.IsSpace(r) }

func skipWS(s string, i int) int {
	for i < len(s) && isWS(rune(s[i])) {
		i++
	}
	return i
}

func shortRest(s string) string {
	s = strings.TrimSpace(s)
	return "`" + truncate(s, 30) + "`"
}

// PassageMode selects how a binding's "passage" is delimited. The spec never
// defines it. See ambiguities.md A-02.
type PassageMode int

const (
	// PassageSincePrevious: the passage runs from the end of the previous
	// recognized narrative binding (or the start of the body) to the start of
	// this one. This implementation's default.
	PassageSincePrevious PassageMode = iota
	// PassageParagraph: the passage is the last blank-line-delimited block
	// before the binding.
	PassageParagraph
	// PassageDocument: the passage is the whole body.
	PassageDocument
)

func passageFor(body string, comments []struct{ start, end int }, idx int, recognized []bool, mode PassageMode) string {
	c := comments[idx]
	switch mode {
	case PassageDocument:
		return body
	case PassageParagraph:
		pre := body[:c.start]
		pre = strings.TrimRight(pre, " \t\n")
		if i := strings.LastIndex(pre, "\n\n"); i >= 0 {
			return pre[i+2:]
		}
		return pre
	default:
		start := 0
		for j := idx - 1; j >= 0; j-- {
			if recognized[j] {
				start = comments[j].end
				break
			}
		}
		return body[start:c.start]
	}
}

func checkNarratives(d *Deployment, c *Corpus, mode PassageMode, rep *Report) {
	for _, lf := range c.Files {
		if lf.Type != "narrative" {
			// A binding-shaped comment in a record body is territory the spec
			// does not describe. See ambiguities.md A-28.
			if lf.Type == "claim" || lf.Type == "survey" {
				for _, cm := range findComments(lf.Body) {
					if isRecognizedBinding(lf.Body[cm.start:cm.end]) {
						rep.Info("ERF-31", lf.Path, lineOf(lf, cm.start), "",
							"a narrative binding appears in the body of a `%s` record; ERF-31 places bindings in narratives and states no rule for one here, so it was recognized but not validated", lf.Type)
					}
				}
			}
			continue
		}
		checkNarrativeDoc(d, c, lf, mode, rep)
	}
}

func lineOf(lf *LoadedFile, off int) int {
	base := lf.BodyLine
	if base == 0 {
		base = 1
	}
	return base + strings.Count(lf.Body[:off], "\n")
}

// ERF-34: "It carries frontmatter with `type: narrative` [...] plus `title`, a
// string; `corpus`, the id of the corpus it belongs to; and `created`, the
// `{timestamp, by}` stamp every other created thing in this format carries."
func checkNarrativeDoc(d *Deployment, c *Corpus, lf *LoadedFile, mode PassageMode, rep *Report) {
	if !lf.HadFrontmatter {
		rep.Violation("ERF-34", lf.Path, 1, "",
			"a narrative carries frontmatter and prose; this file is a bare YAML document with no body")
	}
	for _, k := range lf.FM.Keys {
		if strings.HasPrefix(k, "x_") {
			continue
		}
		if !knownFields["narrative"][k] {
			rep.Violation("ERF-55", lf.Path, lf.FM.Line, "", "field `%s` is not defined for a narrative (ERF-34 names type, title, corpus, created)", k)
		}
	}
	if t := lf.FM.Get("title"); t == nil || !t.IsStr() || strings.TrimSpace(t.Str) == "" {
		rep.Violation("ERF-34", lf.Path, lf.FM.Line, "title", "a narrative MUST carry `title`, a string")
	}
	cid := lf.FM.Get("corpus")
	if cid == nil || !cid.IsStr() || strings.TrimSpace(cid.Str) == "" {
		rep.Violation("ERF-34", lf.Path, lf.FM.Line, "corpus", "a narrative MUST carry `corpus`, the id of the corpus it belongs to")
	} else if !d.corpusDeclared(cid.Str) {
		rep.Violation("ERF-34", lf.Path, cid.Line, "corpus", "`corpus: %s` names no declared corpus in the deployment", cid.Str)
	}
	cr := lf.FM.Get("created")
	if cr == nil {
		rep.Violation("ERF-34", lf.Path, lf.FM.Line, "created", "a narrative MUST carry `created`, the {timestamp, by} stamp")
	} else {
		checkActorStamp(lf, cr, "created", "ERF-34", rep, false)
	}

	comments := findComments(lf.Body)
	recognized := make([]bool, len(comments))
	for i, cm := range comments {
		recognized[i] = isRecognizedBinding(lf.Body[cm.start:cm.end])
	}

	for i, cm := range comments {
		if !recognized[i] {
			continue
		}
		raw := lf.Body[cm.start:cm.end]
		line := lineOf(lf, cm.start)
		b := parseBinding(raw)
		b.File = lf.Path
		b.Line = line
		b.Start, b.End = cm.start, cm.end

		if b.ParseErr != "" {
			// ERF-31: "A binding that does not match this grammar MUST be
			// reported, never skipped."
			rep.Violation("ERF-31", lf.Path, line, "",
				"narrative binding does not match the grammar: %s -- in %s", b.ParseErr, truncate(collapse(raw), 90))
			// The specific case where HTML and the grammar disagree.
			if strings.Contains(raw, "\"") && !strings.HasSuffix(strings.TrimSpace(raw), "-->") {
				rep.Info("ERF-31", lf.Path, line, "",
					"note: an HTML comment ends at the first `-->`, but the ERF-31 grammar lets an anchor contain that sequence; such an anchor is unwritable")
			}
			continue
		}

		b.Passage = passageFor(lf.Body, comments, i, recognized, mode)

		// ERF-31: "The anchor occurs in its passage under ERF-51 [...]
		// A validator MUST flag an anchor that does not occur in its passage."
		if Normalize(b.Anchor) == "" {
			rep.Flag("ERF-31", lf.Path, line, "", "anchor is empty after ERF-51 normalization, so it can pin nothing")
		} else if !strings.Contains(Normalize(b.Passage), Normalize(b.Anchor)) {
			rep.Flag("ERF-31", lf.Path, line, "",
				"anchor %q does not occur in its passage under ERF-51; the prose has probably been edited", truncate(b.Anchor, 70))
		}

		// ERF-33 + ERF-32.
		bound := parseTS(b.BoundAt)
		for _, id := range b.IDs {
			rec, ok := d.ById[id]
			if !ok {
				// ERF-33: "A consumer encountering a narrative binding whose
				// id resolves to no record MUST report it and MUST NOT drop
				// it silently."
				rep.Violation("ERF-33", lf.Path, line, "",
					"narrative binding names `%s`, which resolves to no record in the deployment", id)
				rep.Flag("ERF-32", lf.Path, line, "",
					"staleness of the binding on `%s` is indeterminate: the comparison cannot be run against a record that does not resolve", id)
				continue
			}
			if rec.Type != "claim" {
				// `claims:` names claims. The spec says only "resolves to no
				// record". See ambiguities.md A-28.
				rep.Advisory("ERF-33", lf.Path, line, "",
					"narrative binding names `%s` under `claims:`, but that id is a `%s`, not a claim", id, rec.Type)
			}
			lm := rec.FM.Get("last_modified")
			if lm == nil {
				continue // never edited since minting: not stale
			}
			lmts := stampTS(lm)
			if bound.Precision == TSInvalid || lmts.Precision == TSInvalid {
				rep.Flag("ERF-32", lf.Path, line, "",
					"staleness of the binding on `%s` is indeterminate: a timestamp will not parse (bound-at=%q, last_modified=%q)", id, b.BoundAt, lmts.Raw)
				continue
			}
			switch isStale(bound, lmts) {
			case StaleStale:
				rep.Flag("ERF-32", lf.Path, line, "",
					"narrative binding on `%s` is stale: the claim's last_modified (%s) is later than bound-at=%s", id, lmts.Raw, b.BoundAt)
			case StaleIndeterminate:
				rep.Flag("ERF-32", lf.Path, line, "",
					"narrative binding on `%s` has indeterminate staleness", id)
			}
		}
	}
}

func collapse(s string) string { return strings.Join(strings.Fields(s), " ") }

func stampTS(stamp *Val) TS {
	if stamp == nil {
		return TS{Precision: TSInvalid}
	}
	if stamp.IsMap() {
		return parseTS(stamp.Get("timestamp").StrOr())
	}
	return TS{Precision: TSInvalid}
}

File: reviews/2026-08-25-schema-spec-trials/rust-validator/src/narrative.rs
//! YAMLB-1: the narrative binding's spelling, its recognition, and its grammar.
//!
//! narrative-binding ::= "<!--" ws* "claims:" ws+ ids ws+ anchor
//!                       ws+ "bound-at=" date ws* "-->"

#[derive(Debug)]
pub struct Binding {
    pub ids: Vec<String>,
    pub anchor: String,
    pub bound_at: String,
    /// char offset of `<!--` in the body
    pub start: usize,
    /// char offset just past `-->` in the body
    pub end: usize,
    pub line: usize,
}

#[derive(Debug)]
pub struct BadCandidate {
    pub start: usize,
    pub end: usize,
    pub line: usize,
    pub why: String,
}

#[derive(Debug, Default)]
pub struct Scan {
    pub bindings: Vec<Binding>,
    pub bad: Vec<BadCandidate>,
}

/// Character positions of the body that CommonMark would read as code: fenced
/// code blocks, indented code blocks, and inline code spans. YAMLB-1:
/// "A candidate is `<!--` followed by `claims:` where CommonMark would read an
/// HTML comment, never inside a code span or a code block".
fn code_mask(body: &[char]) -> Vec<bool> {
    let mut mask = vec![false; body.len()];
    // Split into lines, keeping offsets.
    let mut lines: Vec<(usize, usize)> = Vec::new(); // (start, end-exclusive, no newline)
    let mut s = 0usize;
    for (i, &c) in body.iter().enumerate() {
        if c == '\n' {
            lines.push((s, i));
            s = i + 1;
        }
    }
    if s <= body.len() {
        lines.push((s, body.len()));
    }

    let mut fence: Option<(char, usize)> = None; // (char, run length)
    let mut prev_blank = true;
    for &(ls, le) in &lines {
        let line: String = body[ls..le].iter().collect();
        let trimmed_start = line.len() - line.trim_start_matches(' ').len();
        let t = line.trim_start_matches(' ');
        let blank = t.trim().is_empty();

        if let Some((fc, flen)) = fence {
            // inside a fenced block: mask everything, look for the closing fence
            for m in mask.iter_mut().take(le).skip(ls) {
                *m = true;
            }
            let run = t.chars().take_while(|&c| c == fc).count();
            if trimmed_start <= 3 && run >= flen && t.chars().skip(run).all(|c| c == ' ') {
                fence = None;
            }
            prev_blank = blank;
            continue;
        }

        // opening fence?
        if trimmed_start <= 3 {
            for fc in ['`', '~'] {
                let run = t.chars().take_while(|&c| c == fc).count();
                if run >= 3 {
                    fence = Some((fc, run));
                    break;
                }
            }
            if fence.is_some() {
                for m in mask.iter_mut().take(le).skip(ls) {
                    *m = true;
                }
                prev_blank = blank;
                continue;
            }
        }

        // indented code block: >= 4 leading spaces, and not a paragraph continuation
        if trimmed_start >= 4 && prev_blank && !blank {
            for m in mask.iter_mut().take(le).skip(ls) {
                *m = true;
            }
            prev_blank = false;
            continue;
        }

        // inline code spans on this line and the ones after it (CommonMark allows
        // a code span to span lines; this scans within the whole body region
        // between blank lines, approximated per line, which is enough for the
        // hazard YAMLB-1 names).
        mask_code_spans(body, ls, le, &mut mask);
        prev_blank = blank;
    }
    mask
}

fn mask_code_spans(body: &[char], ls: usize, le: usize, mask: &mut [bool]) {
    let mut i = ls;
    while i < le {
        if body[i] == '`' {
            let mut n = 0usize;
            while i + n < le && body[i + n] == '`' {
                n += 1;
            }
            // look for a closing run of exactly n backticks
            let mut j = i + n;
            let mut closed = None;
            while j < le {
                if body[j] == '`' {
                    let mut m = 0usize;
                    while j + m < le && body[j + m] == '`' {
                        m += 1;
                    }
                    if m == n {
                        closed = Some(j + m);
                        break;
                    }
                    j += m;
                } else {
                    j += 1;
                }
            }
            if let Some(endp) = closed {
                for m in mask.iter_mut().take(endp).skip(i) {
                    *m = true;
                }
                i = endp;
                continue;
            }
            i += n;
            continue;
        }
        i += 1;
    }
}

fn line_of(body: &[char], pos: usize) -> usize {
    1 + body[..pos.min(body.len())].iter().filter(|&&c| c == '\n').count()
}

fn starts_with_at(body: &[char], pos: usize, pat: &str) -> bool {
    let p: Vec<char> = pat.chars().collect();
    pos + p.len() <= body.len() && body[pos..pos + p.len()] == p[..]
}

fn find_at(body: &[char], from: usize, pat: &str) -> Option<usize> {
    let p: Vec<char> = pat.chars().collect();
    if p.len() > body.len() {
        return None;
    }
    (from..=body.len().saturating_sub(p.len())).find(|&i| body[i..i + p.len()] == p[..])
}

pub fn scan(body_text: &str) -> Scan {
    let body: Vec<char> = body_text.chars().collect();
    let mask = code_mask(&body);
    let mut out = Scan::default();

    let mut i = 0usize;
    while i < body.len() {
        if !starts_with_at(&body, i, "<!--") || mask[i] {
            i += 1;
            continue;
        }
        // "A candidate is `<!--` followed by `claims:`" (ws* between them).
        let mut k = i + 4;
        while k < body.len() && body[k].is_whitespace() {
            k += 1;
        }
        if !starts_with_at(&body, k, "claims:") {
            // an ordinary HTML comment: skip past it
            i = find_at(&body, i + 4, "-->").map(|p| p + 3).unwrap_or(body.len());
            continue;
        }
        let line = line_of(&body, i);

        // "A candidate is delimited at the first `-->` before the grammar is
        //  applied [...] a candidate whose first `-->` comes after another
        //  `<!--`, or never, is unterminated, extends to the end of its own line".
        let close = find_at(&body, i + 4, "-->");
        let next_open = find_at(&body, i + 4, "<!--");
        let unterminated = match (close, next_open) {
            (None, _) => true,
            (Some(c), Some(o)) => o < c,
            (Some(_), None) => false,
        };
        if unterminated {
            let eol = find_at(&body, i, "\n").unwrap_or(body.len());
            out.bad.push(BadCandidate {
                start: i,
                end: eol,
                line,
                why: "unterminated: the first `-->` never comes, or comes after another `<!--`"
                    .into(),
            });
            i = eol.max(i + 1);
            continue;
        }
        let c = close.unwrap();
        let inner: Vec<char> = body[k + "claims:".len()..c].to_vec();
        match parse_inner(&inner) {
            Ok((ids, anchor, bound_at)) => out.bindings.push(Binding {
                ids,
                anchor,
                bound_at,
                start: i,
                end: c + 3,
                line,
            }),
            Err(why) => out.bad.push(BadCandidate {
                start: i,
                end: c + 3,
                line,
                why,
            }),
        }
        i = c + 3;
    }
    out
}

/// ws+ ids ws+ anchor ws+ "bound-at=" date ws*
fn parse_inner(inner: &[char]) -> Result<(Vec<String>, String, String), String> {
    let mut p = 0usize;
    let ws0 = p;
    while p < inner.len() && inner[p].is_whitespace() {
        p += 1;
    }
    if p == ws0 {
        return Err("no whitespace after `claims:`".into());
    }
    // ids: one or more tokens of characters other than whitespace, '"', '<', '>'.
    let mut ids: Vec<String> = Vec::new();
    loop {
        if p >= inner.len() {
            return Err("the candidate ends before its anchor".into());
        }
        if inner[p] == '"' {
            break;
        }
        let start = p;
        while p < inner.len()
            && !inner[p].is_whitespace()
            && inner[p] != '"'
            && inner[p] != '<'
            && inner[p] != '>'
        {
            p += 1;
        }
        if p == start {
            return Err(format!(
                "the character {:?} is legal in no part of the grammar here",
                inner[p]
            ));
        }
        ids.push(inner[start..p].iter().collect());
        let ws = p;
        while p < inner.len() && inner[p].is_whitespace() {
            p += 1;
        }
        if p == ws && p < inner.len() && inner[p] != '"' {
            return Err("ids must be separated by whitespace".into());
        }
        if p == ws && p < inner.len() && inner[p] == '"' {
            return Err("the anchor must be preceded by whitespace".into());
        }
    }
    if ids.is_empty() {
        return Err("no claim ids before the anchor".into());
    }
    if ids.iter().any(|s| s.contains(',')) {
        return Err(
            "ids are separated by whitespace and never by commas; a comma is part of no id".into(),
        );
    }
    // anchor: '"' char+ '"', with the two escapes \" and \\
    if inner.get(p) != Some(&'"') {
        return Err("no opening quote for the anchor".into());
    }
    p += 1;
    let mut anchor = String::new();
    let mut closed = false;
    while p < inner.len() {
        match inner[p] {
            '"' => {
                closed = true;
                p += 1;
                break;
            }
            '\\' => match inner.get(p + 1) {
                Some('"') => {
                    anchor.push('"');
                    p += 2;
                }
                Some('\\') => {
                    anchor.push('\\');
                    p += 2;
                }
                _ => {
                    return Err(
                        "a backslash in the anchor must open one of the two escapes, \\\" or \\\\"
                            .into(),
                    )
                }
            },
            c => {
                anchor.push(c);
                p += 1;
            }
        }
    }
    if !closed {
        return Err("the anchor is not closed".into());
    }
    if anchor.is_empty() {
        return Err("the anchor is empty; the grammar requires char+".into());
    }
    let ws = p;
    while p < inner.len() && inner[p].is_whitespace() {
        p += 1;
    }
    if p == ws {
        return Err("no whitespace between the anchor and `bound-at=`".into());
    }
    let rest: String = inner[p..].iter().collect();
    let rest_t = rest.trim_end();
    let Some(date) = rest_t.strip_prefix("bound-at=") else {
        return Err(format!(
            "expected `bound-at=` and a date, found {:?}",
            crate::util::truncate(rest_t, 40)
        ));
    };
    if date.len() != 10 || !crate::util::is_date_shape(date) {
        return Err(format!("`bound-at={}` is not YYYY-MM-DD", date));
    }
    Ok((ids, anchor, date.to_string()))
}
